### ISSUE TITLE
Adding Deferred types support

### ISSUE THREAD

**Issue by moufmouf** [thecodingmachine/graphqlite#42]
We should add support for Webonyx Deferred types in order to offer a way to avoid the N+1 issues.

Implementation should be as easy as possible on the user side.

Current proposal:

Consider a request where a user attached to a post must be returned.

```graphql
{
    posts {
        id
        user {
            id
        }
    }
}
```

```php
/**
 * @Type
 */
class PostType {
    /**
     * @Field(prefetchMethod="prefetchUsers")
     * @param Post $post
     * @param mixed $prefetchedUsers
     * @return User
     */
    public function getUser(Post $post, $prefetchedUsers): User
    {
        // This method will receive the $prefetchedUsers as second argument. This is the return value of the "prefetchUsers" method below.
        // Using this prefetched list, it should be easy to map it to the post
    }

    /**
     * @param Post[] $posts
     * @return mixed
     */
    public function prefetchUsers(iterable $posts)
    {
        // This function is called only once per GraphQL request
        // with the list of posts. You can fetch the list of users
        // associated with this posts in a single request,
        // for instance using a "IN" query in SQL or a multi-fetch
        // in your cache back-end.
    }
}
```

When the "prefetchMethod" attribute is detected in the "@Field" annotation, the method is called automatically. The first argument of the method is an iterable over the instances of the main type.
The "prefetchMethod" can return absolutely anything (mixed). The return value will be passed as the second parameter of the "@Field" annotated method.

Important: field arguments can be set either on the @Field annotated method OR/AND on the prefetchMethod.

For instance:

```php
/**
 * @Type
 */
class PostType {
    /**
     * @Field(prefetchMethod="prefetchComments")
     * @param Post $post
     * @param mixed $prefetchedComments
     * @return Comment[]
     */
    public function getComments(Post $post, $prefetchedComments): array
    {
        // ...
    }

    /**
     * @param Post[] $posts
     * @return mixed
     */
    public function prefetchComments(iterable $posts, bool $hideSpam, int $filterByScore)
    {
        // Parameters passed after the first parameter (hideSpam, filterByScore...) are automatically exposed as GraphQL arguments for the "comments" field.
    }
}
```

The prefetch method MUST be in the same class as the @Field-annotated method and MUST be public.

Note: the "prefetchMethod" argument CANNOT be applied on @Query and @Mutation annotations. We will need a different kind of attribute for this.

Note to self: implementation warning! A GraphQL query can contain many different calls with different arguments for a same field. For instance:

```graphql
{
    posts {
        id
        legitComments: comments(hideSpam: true) {
            id
        }
        spamComments: comments(hideSpam: false) {
            id
        }
    }
}
```

With the example above, we will have to call the prefetch method twice! Once with the "hideSpam" set to true, and once with the "hideSpam" set to false.

**Comment by golaod** [comment]
To be honest, this format seems a little bit off. What happened with the original where getComments(Post $post): array would actually be getComments(array $posts): array?

**Comment by golaod** [comment]
Maybe I just don't see exactly how does it work, but let me try though:
```
<?php

/**
 * @Type
 */
class PostType {
    /**
     * @Field(prefetchMethod="eagerLoadComments")
     * @return Comment[]
     */
    public function getComments(Post $post, $promisedData): array
    {
        return array_filter($promisedData, function ($comment) use ($post) {
            return $post->getId() === $comment['post_id'];
        });
    }

    /**
     * @param Post[] $posts
     * @return array
     */
    public function eagerLoadComments(iterable $posts, bool $hideSpam, int $filterByScore)
    {
        return [
            [
                'post_id' => 1,
                'comment' => 'Hello world',
            ],
            [
                'post_id' => 1,
                'comment' => 'Hello world2',
            ],
            [
                'post_id' => 2,
                'comment' => 'Hello world3',
            ],
        ];
    }
}
```

Is this what you wanted to achieve?

**Comment by moufmouf** [comment]
Hey @golaod ,

Your example is perfect, you got it right.

Indeed, the idea we first talked about was different. My first idea was something like:

```php
/**
 * @Type
 */
class PostType {
    /**
     * @BatchField()
     * @param Post[] $posts
     * @return Comment[][]
     */
    public function getComments(array $posts): array
    {
        // ...
    }
}
```

I've been talking with @homersimpsons for quite some time about it. He convinced me this is a somewhat better alternative.

Here are the main points:

**Regarding method signature**

```php
/**
 * @Field(prefetchMethod="eagerLoadComments")
 * @return Comment[]
 */
public function getComments(Post $post, $promisedData): array
```

This method signature is quite similar to the usual way we create fields. It is easier to understand (it is obvious that the field "comments" is of type "Comment[]".

Compare this to:

```php
    /**
     * @BatchField()
     * @param Post[] $posts
     * @return Comment[][]
     */
    public function getComments(array $posts): array
    {
        // ...
    }
```

The return type ("Comment[][]") will map to Comment[] in GraphQL. It is obvious to both of us, but it could cause problems to beginners.

**Regarding implementation**

With this signature:

```
    public function getComments(array $posts): array
```

the keys of the returned array **must match** the keys of the $posts array (otherwise, GraphQLite cannot know by itself what comments are linked to what posts).

Doing the matching on the keys of the arrays is error-prone. Also, what is GraphQLite supposed to return if some keys are missing or if there are some additional keys passed by the developer? Should it trigger an error?

Compare this to the solution proposed by @homersimpsons:

```php
/**
 * @Field(prefetchMethod="eagerLoadComments")
 * @return Comment[]
 */
public function getComments(Post $post, $promisedData): array
```

Here, there is absolutely no possible doubt. The returned comments are matching the "$post" passed in argument. The method signature is way more strict than in the other solution and less things can go awkward.

Does it make sense? Or do you think another way of doing things we did not think about?

**Comment by golaod** [comment]
How does it now about `$promisedData`? Asking because the field can still have inputs

**Comment by moufmouf** [comment]
```php
/**
 * @Field(prefetchMethod="eagerLoadComments")
 * @return Comment[]
 */
public function getComments(Post $post, $promisedData): array
```

GraphQLite will detect the `prefetchMethod` attribute in the @Field annotation. If the `prefetchMethod` attribute is present, the 2nd argument of the method MUST be the promisedData (returned by the prefetch method).

Input arguments come either in third position and afterwards, or in the prefetch method:

```php
/**
 * @Type
 */
class PostType {
    /**
     * @Field(prefetchMethod="prefetchComments")
     * @param Post $post
     * @param mixed $prefetchedComments
     * @return Comment[]
     */
    public function getComments(Post $post, $prefetchedComments, int $maxCommentLength): array
    {
        // In this sample, the input arguments of the comments field are:
        // - maxCommentLength: Int
        // - hideSpam: Bool
        // - filterByScore: Int
    }

    /**
     * @param Post[] $posts
     * @return mixed
     */
    public function prefetchComments(iterable $posts, bool $hideSpam, int $filterByScore)
    {
        // Parameters passed after the first parameter (hideSpam, filterByScore...) are automatically exposed as GraphQL arguments for the "comments" field.
    }
}
```

**Comment by golaod** [comment]
:+1:
